UnitTests/tUnificationWithAbstraction.cpp

```diff
@@ -25,223 +25,290 @@
 #include "Indexing/Index.hpp"
 #include "Indexing/LiteralSubstitutionTree.hpp"
 #include "Indexing/LiteralIndex.hpp"
-
+#include "Indexing/TermSubstitutionTree.hpp"
+#include "Indexing/TermIndex.hpp"
 
 #include "Test/UnitTesting.hpp"
+#include "Test/SyntaxSugar.hpp"
 
 // TODO make this test use assertions, instead of printing output
 
 using namespace Kernel;
 using namespace Indexing;
-using SortType = TermList;
 
-TermList number(vstring n)
-{
-  return TermList(Term::create(env.signature->addIntegerConstant(n,false),0,0));
-}
-TermList var(unsigned i)
+Clause* unit(Literal* lit)
 {
-  return TermList(i,false);
+  static Inference testInf = Kernel::NonspecificInference0(UnitInputType::ASSUMPTION, InferenceRule::INPUT); 
+  Clause * cl = new(1) Clause(1,testInf);
+  (* cl)[0] = lit;
+  return cl;
 }
-unsigned function_symbol(vstring name,unsigned arity,SortType srt)
+
+
+TermIndexingStructure* getTermIndex(bool uwa = true)
 {
-  bool added;
-  unsigned f = env.signature->addFunction(name,arity,added);
-  if(added){
-    Signature::Symbol* symbol = env.signature->getFunction(f);
-    OperatorType* ot = OperatorType::getFunctionTypeUniformRange(arity,srt,srt);
-    symbol->setType(ot); 
+  if(uwa){
+    UWAMismatchHandler* handler = new UWAMismatchHandler();
+    return new TermSubstitutionTree(handler); 
+  } else {
+    HOMismatchHandler* handler = new HOMismatchHandler();
+    return new TermSubstitutionTree(handler);     
   }
-  return f; 
-}
-TermList constant(vstring name,SortType srt)
-{
-  auto c =  function_symbol(name,0,srt);
-  Term* t = Term::create(c,0,0);
-  return TermList(t);
 }
-TermList int_constant(vstring name)
+
+LiteralIndexingStructure* getLiteralIndex()
 {
-  return constant(name,IntegerConstantType::getSort());
+  UWAMismatchHandler* handler = new UWAMismatchHandler();
+  return new LiteralSubstitutionTree(handler); 
 }
-TermList binary(Interpretation fun, TermList n1, TermList n2)
+
+void reportTermMatches(TermIndexingStructure* index, TermList term, TermList sort)
 {
-  return TermList(Term::create2(env.signature->getInterpretingSymbol(fun),n1,n2));
+  TermQueryResultIterator it= index->getUnificationsUsingSorts(term,sort,true);
+  cout << endl;
+  cout << "Unify with " << term.toString() << endl;
+  while(it.hasNext()){
+    TermQueryResult qr = it.next();
+    cout << qr.term.toString() << " matches with substitution: "<< endl;
+    // cout << qr.substitution->tryGetRobSubstitution()->toString() << endl;
+    cout << "and constraints: "<< endl;
+    qr.substitution->numberOfConstraints();
+    auto constraints = qr.substitution->getConstraints();
+    while(constraints.hasNext()){
+      Literal* constraint = constraints.next();
+      cout << "> " << constraint->toString() << endl;
+    }
+  }
+  cout << endl;
 }
```

Comment:
I'm not 100% happy with these unit tests, as they require one to check the stdout of the test case, hence their correctness is not checked by our CI. 
It would be nice to replace the function `reportTermMatches` by `checkTermMatches`, and add an argument `Stack<TermQueryResult> expected`, that is being checked against the iterator.

Anyways this unit test is better than what we have so I'm very happy about having it as is as well!

Reply:
That would be good, but a bit of work. We don't want to make the test dependant on the _order_ that the results are returned.